src/logic/addressBook/store/reducer/addressBook.js

```diff
@@ -98,6 +99,28 @@ export default handleActions<State, *>(
       })
       return newState
     },
+    [ADD_OR_UPDATE_ENTRY]: (state: State, action: ActionType<Function>): State => {
+      const { entry } = action.payload
+
+      // Adds or Updates the entry to all the safes
+      return state.withMutations(map => {
+        const addressBook = map.get('addressBook')
+        if (addressBook) {
+          addressBook.keySeq().forEach(safeAddress => {
+            const safeAddressBook: List<AddressBookEntry> = state.getIn(['addressBook', safeAddress])
+            const entryIndex = safeAddressBook.findIndex(entryItem => sameAddress(entryItem.address, entry.address))
+
+            if (entryIndex !== -1) {
+              const updatedEntriesList = safeAddressBook.set(entryIndex, entry)
```

Comment:
shouldn't this be `.merge`? if we want to update only a single prop for the entry

Reply:
Will check, I just mixed what `ADD_ENTRY` and `UPDATE_ENTRY` were doing

Reply:
Ok, `.setIn` looks like the way to go here. https://stackoverflow.com/a/39788499

Have a look at the changes and give me your feedback: https://github.com/gnosis/safe-react/commit/00a75d15f726dc20fe4c0aba6b2c3bfbe162ef06

Reply:
how would you update a single property of the address book entry given that you have only that prop and an address?

Reply:
What I try to say is that I think update should work with partial data of an entry, and in your approach you have to pass already updated entry

Reply:
Ohh, now I understand your point, and I completely agree that it was an exaggeration to update one prop given that there are only three props for that entry.

Things I'm not getting, and please enlighten me:
- why would I want to not pass the whole object but only an attribute?
- where will I take the entry `address` to find the entry that I need to update in the store?
- what's the advantage of not passing the whole object, if it's small, simple, and typed?

Reply:
Safe model was four props only back in the days :) this is a really very small thing you can do now to save yourself some time later

Reply:
> where will I take the entry address to find the entry that I need to update in the store?

You can pass it as a separate parameter or inside the action payload

Reply:
https://immutable-js.github.io/immutable-js/docs/#/Map/merge

Reply:
Please, be my guess. Because I'm missing the simple solution you're purposing. :pray:

Reply:
Currently, when. you're trying to update an entry, you're overwriting the whole entry (`.set`) with a new one, this is inefficient. We can use `.merge` method from immutable.js to merge current entry with the updated data

Reply:
But to use `.merge` I have to retrieve the _entry_ instead of the `entryIndex`. Merge the object, then `.set` it back again in the `List` index.

I think we're over optimizing here, and it's for one only interaction of the user. This is not part of the constant syncing process we're using in the background.

Probably we can think of a less performant approach that won't affect the final result at all.

Also, I'll roll back the later change I added https://github.com/gnosis/safe-react/pull/688/commits/00a75d15f726dc20fe4c0aba6b2c3bfbe162ef06

Reply:
No, we're not over optimizing, we're doing what the reducer should do when you want to update a property

Reply:
```
if (entryIndex !== -1) {
    const updatedEntriesList = safeAddressBook.update(entryIndex, entry => entry.merge(updatedEntryData))
}
```

Reply:
@mikheevm, I hope changes satisfy your requirements.

- Refactor `addressBookEntry` to be a `Record` https://github.com/gnosis/safe-react/pull/688/commits/8b090c1bc11683ab32c9b1f4e56dddea2b20bd40
- Replace the `set` with `update->merge` flow https://github.com/gnosis/safe-react/pull/688/commits/3ec3a265e34c468c04e48439a3aebcc016d42215
- Refactor `updateOrAddAddressBookEntry` action signature: https://github.com/gnosis/safe-react/pull/688/commits/6107e913bbb37b4a706d871cb25201e2da2d40e7

Looking forward to your feedback

Reply:
LGTM, thanks